src/core/border_router/routing_manager.cpp

```diff
@@ -3110,6 +3183,245 @@ void RoutingManager::RsSender::HandleTimer(void)
+                    continue;
+                }
+
+                // In some cases, the platform may delegate us more than one prefixes, usually one GUA prefix with one
+                // or more ULA prefixes, in this case, we only preserve the smallest one. Since currently GUA prefixes
+                // are allocated from the 2000::/3 block while the ULA prefixes are generated from the fc00::/7 block.
+                // Replace the current selected platform OMR prefix if the one in RA is favored over the current one.
+                if (favoredPrefixFromRa.GetLength() == 0 ||
+                    ((raPrefixPreferredRemaining == 0 || pioPrefix < favoredPrefixFromRa) &&
+                     pio.GetPreferredLifetime() > 0))
+                {
+                    raHasValidPio = true;
+
+                    // From RFC4861: MaxRtrAdvInterval: The maximum time allowed between sending unsolicited multicast
+                    // Router Advertisements from the interface, in seconds.  MUST be no less than 4 seconds and no
+                    // greater than 1800 seconds.
+                    //
+                    // Default: 600 seconds
+                    //
+                    // So it is safe to assume that the platform will generate another RA message for the Thread
+                    // interface within 30 minutes (so the timeout will be refreshed), thus it is safe to set the
+                    // timeout to UINT16_MAX to avoid overflow.
+
+                    favoredPrefixFromRa = pioPrefix;
+                    raPrefixPreferredRemaining =
+                        Min(pio.GetPreferredLifetime(), static_cast<uint32_t>(NumericLimits<uint16_t>::kMax));
+                    raPrefixValidRemaining =
+                        Min(pio.GetValidLifetime(), static_cast<uint32_t>(NumericLimits<uint16_t>::kMax));
+                }
+
+                // Note that the platform may send another RA message to announce that the current prefix we are using
+                // is no longer preferred or no longer valid.
+                if (pioPrefix == GetPrefix())
+                {
+                    raHasValidPio        = true;
+                    currentPrefixUpdated = true;
+                    currentPrefixPreferredRemaining =
+                        Min(pio.GetPreferredLifetime(), static_cast<uint32_t>(NumericLimits<uint16_t>::kMax));
+                    currentPrefixValidRemaining =
+                        Min(pio.GetValidLifetime(), static_cast<uint32_t>(NumericLimits<uint16_t>::kMax));
+                }
+            }
+        }
+
+        VerifyOrExit(raHasValidPio);
+
+        // Evaluate the most favored prefix from RA against the current published one.
+        if (currentPrefixPreferredRemaining == 0)
+        {
+            if (raPrefixPreferredRemaining > 0 || (currentPrefixValidRemaining == 0 && raPrefixValidRemaining > 0))
+            {
+                favoredPrefixFromRa.SetLength(kOmrPrefixLength);
```

Comment:
Is it allowed for the PD prefix to have a different length?

- What if it has a longer length? Is it correct to shorten the length? (say prefix has 92 bit length and we treat it as a 64 bit)
- What if it has a shorter length? Say 32 bit? Is it ok to use it as 64 bit (with zero as new bits)?

Reply:
We have verified before that the prefix should be no longer than than 64

Reply:
> What if it has a longer length? Is it correct to shorten the length? (say prefix has 92 bit length and we treat it as a 64 bit)

We have checked this above. Added a test case to cover this.

> What if it has a shorter length? Say 32 bit? Is it ok to use it as 64 bit (with zero as new bits)?

It is OK to do so, added a test case to cover this.